ext/openssl/ossl_pkey.c

```diff
@@ -128,48 +174,18 @@ ossl_pkey_read_generic(BIO *bio, VALUE pass)
      * we use the different selection as a workaround.
      * https://github.com/openssl/openssl/issues/20657
      */
-    OSSL_DECODER_CTX_free(dctx);
-    dctx = NULL;
-    dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "PEM", NULL, NULL,
-                                         EVP_PKEY_KEYPAIR, NULL, NULL);
-    if (!dctx)
-        goto out;
-    if (OSSL_DECODER_CTX_set_pem_password_cb(dctx, ossl_pem_passwd_cb, ppass) != 1)
+    pkey = ossl_pkey_read(bio, "PEM", EVP_PKEY_KEYPAIR, pass);
+    if (pkey) {
         goto out;
-    while (1) {
-        if (OSSL_DECODER_from_bio(dctx, bio) == 1)
-            goto out;
-        if (BIO_eof(bio))
-            break;
-        pos2 = BIO_tell(bio);
-        if (pos2 < 0 || pos2 <= pos)
-            break;
-        ossl_clear_error();
-        pos = pos2;
     }
 
-    OSSL_BIO_reset(bio);
-    OSSL_DECODER_CTX_free(dctx);
-    dctx = NULL;
-    dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "PEM", NULL, NULL, 0, NULL, NULL);
-    if (!dctx)
-        goto out;
-    if (OSSL_DECODER_CTX_set_pem_password_cb(dctx, ossl_pem_passwd_cb, ppass) != 1)
-        goto out;
-    while (1) {
-        if (OSSL_DECODER_from_bio(dctx, bio) == 1)
+    for (i = 0; i < selection_num; i++) {
+        pkey = ossl_pkey_read(bio, "PEM", selections[i], pass);
+        if (pkey) {
             goto out;
-        if (BIO_eof(bio))
-            break;
-        pos2 = BIO_tell(bio);
-        if (pos2 < 0 || pos2 <= pos)
-            break;
-        ossl_clear_error();
-        pos = pos2;
+        }
```

Comment:
It seems a bit wasteful that we now call `ossl_pkey_read(bio, "PEM", EVP_PKEY_KEYPAIR, pass)` twice.

Is the order of selections important?

Reply:
Ah, that's good point. I don't think the order of the selection is important. You meant if we call the selections calling `EVP_PKEY_KEYPAIR` first like this, we can remove the following part (`ossl_pkey_read(bio, "PEM", EVP_PKEY_KEYPAIR, pass)`), right?

```
    int selections[] = {
        EVP_PKEY_KEYPAIR,
        EVP_PKEY_KEY_PARAMETERS,
        EVP_PKEY_PUBLIC_KEY
    };
```

```
    pkey = ossl_pkey_read(bio, "PEM", EVP_PKEY_KEYPAIR, pass);
    if (pkey) {
        goto out; 
    }
```

Reply:
I was wondering if there could be a string that can be both decoded as a DER encoding of a private key and also a public key (of a different algorithm).

But looking at the OpenSSL <= 1.1 version of `ossl_pkey_read_generic()`, I guess this should be OK. We used to do exactly the same thing: first try to decode as a private key, then as a public key.

Reply:
> I was wondering if there could be a string that can be both decoded as a DER encoding of a private key and also a public key (of a different algorithm).

OK. I guess that you are talking about a possibility that there could be `EVP_PKEY_KEY_*` that "can be both decoded as a DER encoding of a private key and also a public key (of a different algorithm)". I am not familiar with the meaning of each `EVP_PKEY_KEY_*`. I just put the meanings of all the `EVP_PKEY_KEY_*` for the record.

```
$ man 3 EVP_PKEY_fromdata
...
   Selections
       The following constants can be used for selection:

       EVP_PKEY_KEY_PARAMETERS
           Only key parameters will be selected.

       EVP_PKEY_PUBLIC_KEY
           Only public key components will be selected. This includes optional key
           parameters.

       EVP_PKEY_KEYPAIR
           Any keypair components will be selected. This includes the private key,
           public key and key parameters.
...
```